### PR TITLE
Fixes for HIP detection for recent ROCm and CMake

### PR DESCRIPTION
**Pull request by havogt**
- New minimum CMake version requirement 3.21.0 for HIP language support
- Hip detection via `check_language` instead of compiler binary name
- Enable workarounds for Clang 17 (which where already needed for Clang 16 and earlier)
- Fix for deprecated `ptrAttributes.memoryType` (was already changed in CUDA since a while)
- Add cuda error checks for functions that return an error with `[[no_discard]]` in hip.

### CODE REVIEW

**Comment by havogt** [comment]
launch jenkins

**Comment by havogt** [comment]
launch jenkins

File: cmake/public/gridtools_setup_targets.cmake [hunk 1]
```diff
@@ -161,10 +161,15 @@ macro(_gt_setup_targets _config_mode clang_cuda_mode)
...
-        if((GT_CUDA_TYPE STREQUAL NVCC-CUDA) AND (CMAKE_PROJECT_NAME STREQUAL PROJECT_NAME))
+        if(CMAKE_PROJECT_NAME STREQUAL PROJECT_NAME)
             # Do not enable the language if we are included from a super-project.
-            # It is up to the super-project to enable CUDA.
-            enable_language(CUDA)
+            # It is up to the super-project to enable CUDA/HIP.
+            if(GT_CUDA_TYPE STREQUAL NVCC-CUDA)
+                enable_language(CUDA)
+            endif()
+            if(GT_CUDA_TYPE STREQUAL HIPCC-AMDGPU)
```
Comment (fthaler) [inline cmake/public/gridtools_setup_targets.cmake:170]:
```suggestion
            elseif(GT_CUDA_TYPE STREQUAL HIPCC-AMDGPU)
```

File: include/gridtools/common/cuda_util.hpp [hunk 2]
```diff
@@ -37,7 +37,9 @@ namespace gridtools {
             struct cuda_free {
                 template <class T>
                 void operator()(T *ptr) const {
-                    cudaFree(const_cast<std::remove_cv_t<T> *>(ptr));
+                    cudaError_t err = cudaFree(const_cast<std::remove_cv_t<T> *>(ptr));
+                    if (err != cudaSuccess)
+                        on_error(err, "cudaFree", "", "", 0);
```
Comment (fthaler) [inline include/gridtools/common/cuda_util.hpp:42]: Why not `GT_CUDA_CHECK`?
```suggestion
                    GT_CUDA_CHECK(cudaFree(const_cast<std::remove_cv_t<T> *>(ptr)));
```

**Review by fthaler — CHANGES_REQUESTED** [review]

File: CMakeLists.txt [hunk 3]
```diff
@@ -29,12 +29,13 @@ list(APPEND CMAKE_MODULE_PATH "${CMAKE_CURRENT_LIST_DIR}/cmake/public" "${CMAKE_
 set(CMAKE_EXPORT_NO_PACKAGE_REGISTRY ON CACHE BOOL "")
 mark_as_advanced(CMAKE_EXPORT_NO_PACKAGE_REGISTRY)
 
-# User setting GT_CLANG_CUDA_MODE: decide if Clang-CUDA or NVCC
+# User setting GT_CLANG_CUDA_MODE: decide if Clang-CUDA, NVCC or HIP
+# TODO(havogt): rename variable to GT_CLANG_MODE in backwards compatible way
```
[the remark this answers is not in the document]
Comment (havogt) [reply CMakeLists.txt:33]: I would leave that TODO for now.

**Comment by havogt** [comment]
launch jenkins

**Review by fthaler — APPROVED** [review]

**Comment by havogt** [comment]
launch jenkins

**Pull request merged**